app/models/contribution.rb

```diff
@@ -39,13 +55,17 @@ class Contribution
 
   delegate :to_rdf, to: :ore_aggregation
 
+  after_save :set_oai_pmh_fields, if: :published?
```

Comment:
if these are only to be set once on publication, why not move the call into a after block on the publish event?